snforge_std/src/event.cairo

```diff
@@ -0,0 +1,205 @@
+
+        let mut i = 0;
+        loop {
+            if i >= events.len() {
+                break;
+            }
+            self.events.append(events.at(i).clone());
+            i += 1;
+        }
+    }
+}
+
+trait EventAssertions {
+    fn assert_emitted(ref self: EventSpy, events: @Array<Event>);
+}
+
+impl EventAssertionsImpl of EventAssertions {
+    fn assert_emitted(ref self: EventSpy, events: @Array<Event>) {
+        self.fetch_events();
+
+        let mut i = 0;
+        loop {
+            if i >= events.len() {
+                break;
+            }
+
+            let emitted = check_if_emitted(ref self, copy_event(events.at(i)));
+
+            if !emitted {
+                panic(
+                    array![
+                        *events.at(i).name,
+                        'event was not emitted from',
+                        (*events.at(i).from).into()
+                    ]
+                );
+            }
+
+            i += 1;
+        };
+    }
+}
+
+fn check_if_emitted(ref self: EventSpy, event: Event) -> bool {
+    let emitted_events = @self.events;
+    let mut j = 0;
+    return loop {
+        if j >= emitted_events.len() {
+            break false;
+        }
+
+        if event_name_hash(event.name) == *self.events.at(j).name
+            && event.from == *emitted_events.at(j).from {
+            if @event.keys != emitted_events.at(j).keys
+                || @event.data != emitted_events.at(j).data {
+                panic(
+                    array![
+                        event.name,
+                        'event was emitted from',
+                        event.from.into(),
+                        'but keys or data are different'
+                    ]
+                );
+            }
+
+            remove_event(ref self, j);
+
+            break true;
+        }
+
+        j += 1;
+    };
+}
+
+fn remove_event(ref self: EventSpy, index: usize) {
+    let emitted_events = @self.events;
+    let mut emitted_events_deleted_event = array![];
+    let mut k = 0;
+    loop {
+        if k >= emitted_events.len() {
+            break;
+        }
+
+        if k != index {
+            emitted_events_deleted_event.append(copy_event(emitted_events.at(k)));
+        }
+        k += 1;
+    };
+    self.events = emitted_events_deleted_event;
+}
+
+fn copy_event(event: @Event) -> Event {
+    let from = *event.from;
+    let name = *event.name;
+
+    let mut keys = array![];
+    let mut i = 0;
+    loop {
+        if i >= event.keys.len() {
+            break;
+        }
+
+        keys.append(*event.keys.at(i));
+        i += 1;
+    };
+
+    let mut data = array![];
+    i = 0;
+    loop {
+        if i >= event.data.len() {
+            break;
+        }
+
+        data.append(*event.data.at(i));
+        i += 1;
+    };
+
+    Event { from, name, keys, data }
+}
+
+// Copied from corelib until it will be included in the release
```

Comment:
This should be tracked by an issue i think, and we should probably ask the compiler team if it's going to be included

Reply:
Afaik it will be included in the 2.2.0 release

Reply:
#509